Sources/Account.swift

```diff
@@ -7,35 +7,153 @@
+        guard let wallet = wallet else {
+            fatalError("Wallet no longer exists")
+        }
+
+        let key = wallet.key
+        switch key.type {
+        case .encryptedKey:
+            var key = try key.decrypt(password: password)
+            defer {
+                // Clear memory after signing
+                key.resetBytes(in: 0..<key.count)
+            }
+            return PrivateKey(data: key)!
+        case .hierarchicalDeterministicWallet:
+            guard var mnemonic = String(data: try key.decrypt(password: password), encoding: .ascii) else {
+                throw DecryptError.invalidPassword
+            }
+            defer {
+                // Clear memory after signing
+                mnemonic.clear()
+            }
+            let wallet = HDWallet(mnemonic: mnemonic, passphrase: key.passphrase)
+            return wallet.getKey(at: derivationPath)
+        }
+    }
+
+    // MARK: Hashable
+
     public var hashValue: Int {
-        return address.hashValue
+        return address.data.hashValue
     }
 
     public static func == (lhs: Account, rhs: Account) -> Bool {
-        return lhs.address == rhs.address
+        return lhs.address.data == rhs.address.data
+    }
+
+    // MARK: Codable
+
+    enum CodingKeys: String, CodingKey {
+        case blockchain
+        case addressData
+        case derivationPath
+    }
+
+    public init(from decoder: Decoder) throws {
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+        let blockchain = try container.decode(Blockchain.self, forKey: .blockchain)
+        let addressData = try container.decode(Data.self, forKey: .addressData)
+
+        let maybeAddress: Address?
+        switch blockchain {
+        case .bitcoin:
+            maybeAddress = BitcoinAddress(data: addressData)
+        case .ethereum:
+            maybeAddress = EthereumAddress(data: addressData)
+        }
+
+        guard let address = maybeAddress else {
+            throw DecodingError.dataCorruptedError(forKey: .addressData, in: container, debugDescription: "Invalid address")
+        }
+
+        self.address = address
+        derivationPath = try container.decode(DerivationPath.self, forKey: .derivationPath)
+    }
+
+    public func encode(to encoder: Encoder) throws {
+        var container = encoder.container(keyedBy: CodingKeys.self)
+        try container.encode(address.data, forKey: .addressData)
+        try container.encode(address.blockchain, forKey: .blockchain)
+        try container.encode(derivationPath, forKey: .derivationPath)
     }
 }
 
-/// Support account types.
-public enum AccountType {
-    case encryptedKey
-    case hierarchicalDeterministicWallet
+extension Blockchain: Codable {
+    public init(from decoder: Decoder) throws {
+        let container = try decoder.singleValueContainer()
+        let string = try container.decode(String.self)
+        switch string {
+        case "bitcoin":
```

Comment:
maybe adding `bitcoin` `ethereum` as `var id: String` to `Blockchain`?

so that then `Blockchain.ethereum.id`